File: dart/lib/src/utils/custom_bg.dart
import '../constants/addresses.dart';
import '../errors/sns_errors.dart';
import '../types/custom_bg.dart';
import 'get_domain_key_sync.dart';

/// Artist public keys for custom backgrounds
const String _degenPoetKey = 'ART5dr4bDic2sQVZoFheEmUxwQq5VGSx9he7JxHcXNQD';
const String _rgb0x00Key = 'CSWvuDHXExVGEMR9kP8xYAHuNjXogeRck9Cnr312CC9g';
const String _retardioKey = 'J2Q2j6kpSg7tq8JzueCHNTQNcyNnQkvr85RhsFnYZWeG';
const String _numberArtKey = '6vwnZJZNQjtY4zR93YUuyeDUBhacLLH2mQaZiJAvVwzu';

/// Result of custom background key derivation
class CustomBgKeys {
  const CustomBgKeys({
    required this.domainKey,
    required this.bgKey,
  });

  /// The domain key for the custom background
  final String domainKey;

  /// The background key
  final String bgKey;
}

/// Gets the custom background keys for a domain
///
/// This function mirrors js/src/custom-bg.ts getCustomBgKeys
///
/// [domain] - The domain name
/// [customBg] - The custom background type
///
/// Returns the domain and background keys
Future<CustomBgKeys> getCustomBgKeys(String domain, CustomBg customBg) async {
  final hashedBg = getHashedNameSync(customBg.value);
  final hashedDomain = getHashedNameSync(domain);

  final domainKey = await getNameAccountKeySync(
    hashedDomain,
    nameParent: customBgTld,
  );

  final bgKey = await getNameAccountKeySync(
    hashedBg,
    nameParent: domainKey,
  );

  return CustomBgKeys(
    domainKey: domainKey,
    bgKey: bgKey,
  );
}

/// Gets the artist public key for a custom background
///
/// This function mirrors js/src/custom-bg.ts getArtistPubkey
///
/// [bg] - The custom background type
///
/// Returns the artist's public key
/// Throws [InvalidCustomBgError] if the background is invalid
String getArtistPubkey(CustomBg bg) {
  switch (bg) {
    case CustomBg.degenPoet1:
      return _degenPoetKey;
    case CustomBg.rgb0x001:
      return _rgb0x00Key;
    case CustomBg.retardio1:
    case CustomBg.retardio2:
    case CustomBg.retardio3:
      return _retardioKey;
    case CustomBg.numberArt0:
    case CustomBg.numberArt1:
    case CustomBg.numberArt2:
    case CustomBg.numberArt3:
    case CustomBg.numberArt4:
    case CustomBg.numberArt5:
    case CustomBg.numberArt6:
    case CustomBg.numberArt7:
    case CustomBg.numberArt8:
    case CustomBg.numberArt9:
      return _numberArtKey;
    case CustomBg.valentineDay2025:
    case CustomBg.monkedao:
      return vaultOwner;
  }
}

File: js/src/custom-bg.ts
import { PublicKey } from "@solana/web3.js";
import { CUSTOM_BG_TLD, VAULT_OWNER } from "./constants";
import { CustomBg } from "./types/custom-bg";
import { getHashedNameSync } from "./utils/getHashedNameSync";
import { getNameAccountKeySync } from "./utils/getNameAccountKeySync";
import { InvalidCustomBgError } from "./error";

const DEGEN_POET_KEY = new PublicKey(
  "ART5dr4bDic2sQVZoFheEmUxwQq5VGSx9he7JxHcXNQD",
);
const RBG_0x00_KEY = new PublicKey(
  "CSWvuDHXExVGEMR9kP8xYAHuNjXogeRck9Cnr312CC9g",
);
const RETARDIO_KEY = new PublicKey(
  "J2Q2j6kpSg7tq8JzueCHNTQNcyNnQkvr85RhsFnYZWeG",
);
const NUMBER_ART_KEY = new PublicKey(
  "6vwnZJZNQjtY4zR93YUuyeDUBhacLLH2mQaZiJAvVwzu",
);

/**
 * Derives the name account keys for a custom background.
 *
 * @param domain Domain name with its TLD suffix trimmed
 * @param customBg Custom background identifier
 * @returns Custom background domain key and background entry key.
 */
export const getCustomBgKeys = (domain: string, customBg: CustomBg) => {
  const hashedBg = getHashedNameSync(customBg);
  const hashedDomain = getHashedNameSync(domain);

  const domainKey = getNameAccountKeySync(
    hashedDomain,
    undefined,
    CUSTOM_BG_TLD,
  );
  const bgKey = getNameAccountKeySync(hashedBg, undefined, domainKey);

  return { domainKey, bgKey };
};

/**
 * Returns the public key associated with a custom background.
 *
 * @param bg The custom background identifier
 * @returns The artist or payout public key for the background
 */
export const getArtistPubkey = (bg: CustomBg): PublicKey => {
  switch (bg) {
    case CustomBg.DegenPoet1:
      return DEGEN_POET_KEY;
    case CustomBg.Rgb0x001:
      return RBG_0x00_KEY;
    case CustomBg.Retardio1:
      return RETARDIO_KEY;
    case CustomBg.Retardio2:
      return RETARDIO_KEY;
    case CustomBg.Retardio3:
      return RETARDIO_KEY;
    case CustomBg.NumberArt0:
    case CustomBg.NumberArt1:
    case CustomBg.NumberArt2:
    case CustomBg.NumberArt3:
    case CustomBg.NumberArt4:
    case CustomBg.NumberArt5:
    case CustomBg.NumberArt6:
    case CustomBg.NumberArt7:
    case CustomBg.NumberArt8:
    case CustomBg.NumberArt9:
      return NUMBER_ART_KEY;
    case CustomBg.ValentineDay2025:
    case CustomBg.Monkedao:
    case CustomBg.LunarNewYear2026:
    case CustomBg.WorldCup2026:
      return VAULT_OWNER;
    default:
      throw new InvalidCustomBgError("The given background is invalid");
  }
};
